### PR TITLE
Run end-to-end tests in Espresso mode

### LINKED ISSUE #7: Enable preconfirmations
Fork op-geth and modify validation mode to fetch new blocks from sequencer query service instead of L1, ie: enable preconfirmations

### LINKED ISSUE #8: Enable bridge

### LINKED ISSUE #30: Implement `espresso.Client`
It is mocked out for now, as we can make the sequencer chaos monkey tests work without a real client, but we will need this to run the real sequencer

### LINKED ISSUE #31: Add an e2e test for the op-node in Espresso mode
Port the following tests to check both legacy and Espresso mode:
- [x] `TestSystemE2E` in `system_test.go` (basic deposit test)
- [x] `TestConfirmationDepth` in `system_test.go` (tests sequencer and verifier with non-default config)
- [x] `TestMintOnRevertedDeposit` in `system_test.go` (tests deposit transactions)
- [ ] ~~`TestSystemMockP2P` in `system_test.go` (tests preconfirmations)~~ (skipped due to being flaky in upstream CI)
- [x] `TestSystemRPCAltSync` in `system_test.go` (tests preconfirmations using RPC sync instead of gossip)
- [x] `TestSystemP2PAltSync` (tests preconfirmations using p2p gossibp instead of RPC sync)
- [x] `TestWithdrawals` (tests ETH bridge)
- [x] `TestERC20BridgeDeposits` in `bridge_test.go` (tests ERC-20 token bridge)

IMO it is not necessary to port the l2_sequencer tests or add new ones as
* they are whitebox and won't work with Espresso mode without substantial changes
* the e2e tests listed above thoroughly test the sequencer, including submitting transactions to L2 and deposits from L1

### LINKED ISSUE #43: Test deposits with Espresso mode
In theory, the design of our integration should preserve the functionality of the L1->L2 bridge (unlike our Polygon integration which broke it). We should test this to see if it really does.

### PR DESCRIPTION
**Pull request by jbearer**
Depends on https://github.com/EspressoSystems/espresso-sequencer/pull/570
Depends on https://github.com/EspressoSystems/op-geth/pull/2

Closes #31 
Closes #30 
Closes #43 
Closes #8 
Closes #7

### CODE REVIEW

**Comment by jbearer** [comment]
Adding the `espresso.Client` implementation to this PR since I need it to run the tests and I need the tests to check the client

**Comment by jbearer** [comment]
Current status: I'm running an end-to-end test in Espresso mode with

```bash
cd op-e2e
OP_E2E_USE_HTTP=true go test -v \
                -l1-allocs ../.devnet-espresso/allocs-l1.json \
                -l1-deployments ../.devnet-espresso/addresses.json \
                -deploy-config ../packages/contracts-bedrock/deploy-config/devnetL1-espresso.json \
                -run TestSystemE2E ./... 2>&1 | tee /tmp/log.txt
```

The first thing this test does is force a deposit through L1, without sending it to the sequencer. This actually works, which is awesome! It means the L1-L2 bridge/escape hatch is going to work after this integration.

Next it sequences an L2 transaction by sending it to the sequencer RPC (which I have configured to point at the op-geth-proxy service). From logs, I can see that the transaction is sequenced by HotShot and included in a batch by the OP sequencer, which is then sent to op-geth and sealed. Also awesome!

However, the test gets stuck immediately after that, because the next HotShot block includes the same transaction, whose nonce is of course invalid the second time. op-geth refuses to execute a block with an invalid transaction, and the sequencer gets stuck in a loop of trying to seal the same batch repeatedly until the test times out. So, it is looking like we'll need to do something about invalid transactions :(

**Comment by jbearer** [comment]
Found the place where they add an exception for deposit transactions, where invalid transactions are converted to failed transactions and still added to the block: https://github.com/ethereum-optimism/op-geth/blob/optimism/core/state_transition.go#L355.

We'll have to fork op-geth, but I think adding a similar exception for Espresso transactions should fix the problem. It makes sense that these would be treated similarly, as they both have the same problem: they are force-included (deposit by L1, Espresso transactions by HotShot), the sequencer has no discretion to discard invalid ones.

**Comment by nomaxg** [comment]
> Found the place where they add an exception for deposit transactions, where invalid transactions are converted to failed transactions and still added to the block: https://github.com/ethereum-optimism/op-geth/blob/optimism/core/state_transition.go#L355.
> 
> We'll have to fork op-geth, but I think adding a similar exception for Espresso transactions should fix the problem. It makes sense that these would be treated similarly, as they both have the same problem: they are force-included (deposit by L1, Espresso transactions by HotShot), the sequencer has no discretion to discard invalid ones.

We could also have the sequencer remove failed transactions from the payload by repeatedly shrinking the txns payload until the engine succeeds and somehow submit failed txn indices to the batch inbox so that validators can deterministically remove the same transactions before executing. However, this adds a lot of overhead to the pipeline so forking op-geth may be the better approach.

**Comment by jbearer** [comment]
Yeah, I think of this as a last resort. The derivation pipeline would also have to do the same thing, execute the full batch and gradually shrink it, to determine that the sequencer shrank it honestly.

I don't think it will be _too_ much work to at least see if patching op-geth is feasible. I'll try to find out today.

**Comment by jbearer** [comment]
A bunch of tests are now passing. Many are not, but I think at this point the best thing to do is clean up CI and then merge it, so we can move on to working on the demo and cleaning up some loose ends. We can selectively fix and enable more tests time permitting. I suspect many of the failing tests are just having timeouts, since there's a little bit of extra latency in Espresso mode compared to the centralized sequencer.

File: go.mod [hunk 1]
```diff
@@ -198,6 +199,6 @@ require (
 	rsc.io/tmplfunc v0.0.3 // indirect
 )
 
-replace github.com/ethereum/go-ethereum v1.12.0 => github.com/ethereum-optimism/op-geth v1.101106.1-0.20230724181546-b9c6d36ae9b8
+replace github.com/ethereum/go-ethereum => github.com/EspressoSystems/op-geth v1.2.0-espresso.0.20230829005208-4111f649c559
 
-//replace github.com/ethereum/go-ethereum v1.12.0 => ../go-ethereum
+//replace github.com/ethereum/go-ethereum v1.12.0 => ../op-geth
```
Comment (nomaxg) [inline go.mod]: remove this comment?

File: op-e2e/setup.go [hunk 2]
```diff
@@ -236,12 +246,148 @@ func (sys *System) Close() {
...
+		select {
+		case <-ticker.C:
+			continue
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+func (e *EspressoSystem) StartGethProxy(sequencer *node.Node) error {
```
Comment (nomaxg) [inline op-e2e/setup.go:306]: Nice, we should add a real hotshot commitment task/contract deployment to this system eventually

File: op-e2e/setup.go [hunk 3]
```diff
@@ -236,12 +246,148 @@ func (sys *System) Close() {
...
+		if err := e.logsProcess.Process.Kill(); err != nil {
+			log.Error("failed to kill docker-compose logs", "err", err)
+		}
+		if err := e.logsProcess.Wait(); err != nil {
+			log.Error("failed to wait for docker-compose logs", "err", err)
+		}
+	}
+}
+
+func dockerComposePort(projectName string, composeFile string, service string, internalPort uint16) (uint16, error) {
```
Comment (nomaxg) [inline op-e2e/setup.go:374]: nit: DockerComposePort

**Review by nomaxg — APPROVED** [review]
LGTM

File: op-e2e/setup.go (hunk 3 above)
Comment (jbearer) [reply op-e2e/setup.go:374]: I think lower case is correct because it's private to this file? At least that is what I have gathered is the convention

**Comment by jbearer** [comment]
I'm pretty sure the failing test is just flaky. It passes sometimes, and when it fails it's just a timeout. And it passes consistently for me locally. Since it's a legacy mode test I don't think it's worth debugging right now, I'm just going to merge this

**Pull request merged**